cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -0,0 +1,139 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# Accumulate all dependencies to provide suitable static link parameters to the
+# third party libraries.
+set(ICEBERG_SYSTEM_DEPENDENCIES)
+set(ICEBERG_VENDOR_DEPENDENCIES)
+set(ICEBERG_ARROW_INSTALL_INTERFACE_LIBS)
+
+# ----------------------------------------------------------------------
+# Versions and URLs for toolchain builds
+
+set(ICEBERG_ARROW_BUILD_VERSION "18.1.0")
+set(ICEBERG_ARROW_BUILD_SHA256_CHECKSUM
+    "2dc8da5f8796afe213ecc5e5aba85bb82d91520eff3cf315784a52d0fa61d7fc")
+set(ARROW_VENDORED TRUE)
+
+if(DEFINED ENV{ICEBERG_ARROW_URL})
+  set(ARROW_SOURCE_URL "$ENV{ICEBERG_ARROW_URL}")
+else()
+  set(ARROW_SOURCE_URL
+      "https://www.apache.org/dyn/closer.cgi?action=download&filename=/arrow/arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+      "https://downloads.apache.org/arrow/arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+      "https://github.com/apache/arrow/releases/download/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}/apache-arrow-${ICEBERG_ARROW_BUILD_VERSION}.tar.gz"
+  )
+endif()
+
+# ----------------------------------------------------------------------
+# FetchContent
+
+include(FetchContent)
+set(FC_DECLARE_COMMON_OPTIONS)
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.28)
+  list(APPEND FC_DECLARE_COMMON_OPTIONS EXCLUDE_FROM_ALL TRUE)
+endif()
+
+macro(prepare_fetchcontent)
+  set(BUILD_SHARED_LIBS OFF)
+  set(BUILD_STATIC_LIBS ON)
+  set(CMAKE_COMPILE_WARNING_AS_ERROR FALSE)
+  set(CMAKE_EXPORT_NO_PACKAGE_REGISTRY TRUE)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+endmacro()
+
+# ----------------------------------------------------------------------
+# Apache Arrow
+
+function(resolve_arrow_dependency)
+  prepare_fetchcontent()
+
+  set(ARROW_BUILD_SHARED
+      OFF
+      CACHE BOOL "" FORCE)
+  set(ARROW_BUILD_STATIC
+      ON
+      CACHE BOOL "" FORCE)
+  set(ARROW_FILESYSTEM
+      OFF
+      CACHE BOOL "" FORCE)
+  set(ARROW_SIMD_LEVEL
+      "NONE"
+      CACHE STRING "" FORCE)
+  set(ARROW_RUNTIME_SIMD_LEVEL
+      "NONE"
+      CACHE STRING "" FORCE)
+  set(ARROW_POSITION_INDEPENDENT_CODE
+      ON
+      CACHE BOOL "" FORCE)
+
+  fetchcontent_declare(Arrow
```

Comment:
@kou Sorry to bother you. I tried to add a minimal dependency on Apache Arrow to this project. The CI fails only on Windows:
```
CMake Error at build/_deps/arrow-src/cpp/src/arrow/CMakeLists.txt:587 (target_link_libraries):
  Target "arrow_util_static" links to:

    Boost::headers

CMake Error at build/_deps/arrow-src/cpp/src/arrow/CMakeLists.txt:804 (target_link_libraries):
  Target "arrow_compute_static" links to:

    Boost::headers

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

It seems that Boost is enabled due to missing native support of INT128 but not built via resolve_dependency.
```
-- Looking for _M_ARM64
-- Looking for _M_ARM64 - not found
-- Looking for __SIZEOF_INT128__
-- Looking for __SIZEOF_INT128__ - not found
-- Boost include dir: 
```

Reply:
OK, I got it compiled. The default values of Arrow `define_string_option are somehow missing so I have to explicitly set ARROW_DEPENDENCY_SOURCE to AUTO/BUNDLED to build Boost. Similar for ARROW_SIMD_LEVEL and ARROW_RUNTIME_SIMD_LEVEL.

Reply:
Yes, I know that I cannot simply override Apache Arrow's third-party dependencies at this moment.